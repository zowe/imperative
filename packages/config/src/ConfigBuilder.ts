/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import * as lodash from "lodash";
import { IImperativeConfig } from "../../imperative";
import { Config } from "./Config";
import { IConfig } from "./doc/IConfig";
import { IConfigBuilderOpts } from "./doc/IConfigBuilderOpts";
import { IConfigProfile } from "./doc/IConfigProfile";

export class ConfigBuilder {
    /**
     * Build a new Config object from an Imperative CLI app configuration.
     * @param impConfig The Imperative CLI app configuration.
     * @param opts Options to control aspects of the builder.
     */
    public static async build(impConfig: IImperativeConfig, opts?: IConfigBuilderOpts): Promise<IConfig> {
        opts = opts || {};
        const config: IConfig = Config.empty();

        for (const profile of impConfig.profiles) {
            const properties: { [key: string]: any } = {};
            const secureProps: string[] = [];
            for (const [k, v] of Object.entries(profile.schema.properties)) {
                if (opts.populateProperties && v.includeInTemplate) {
                    if (v.secure) {
                        secureProps.push(k);
                        const propValue = opts.getSecureValue ? await opts.getSecureValue(k, v) : null;
                        if (propValue != null) {
                            // Save this value to be stored securely after profile is built
                            properties[k] = propValue;
                        }
                    } else {
                        if ((v as any).optionDefinition != null) {
                            // Use default value of ICommandOptionDefinition if present
                            properties[k] = (v as any).optionDefinition.defaultValue;
                        }
                        if (properties[k] === undefined) {
                            // Fall back to an empty value
                            properties[k] = this.getDefaultValue(v.type);
                        }
                    }
                }
            }

            // Add the profile to config and set it as default
            lodash.set(config, `profiles.${profile.type}`, {
                type: profile.type,
                properties,
                secure: secureProps
            });

            if (opts.populateProperties) {
                config.defaults[profile.type] = profile.type;
            }
        }

        // Hoist duplicate default properties
        if (config.profiles != null && impConfig.baseProfile != null) {
            this.hoistTemplateProperties(config.profiles, impConfig.baseProfile.type);
        }

        return { ...config, autoStore: true };
    }

    /**
     * Returns empty value that is appropriate for the property type.
     * @param propType The type of profile property
     * @returns Null or empty object
     */
    private static getDefaultValue(propType: string | string[]): any {
        // TODO How to handle profile property with multiple types
        if (Array.isArray(propType)) {
            propType = propType[0];
        }
        switch (propType) {
            case "string":  return "";
            case "number":  return 0;
            case "object":  return {};
            case "array":   return [];
            case "boolean": return false;
            default:        return null;
        }
    }

    /**
     * Moves duplicate default properties across child profiles into a base
     * profile.
     * @param profiles The profiles object of the config JSON
     * @param baseProfileType The type of the base profile
     */
    private static hoistTemplateProperties(profiles: { [key: string]: IConfigProfile }, baseProfileType: string): void {
        // Flatten properties into object that maps property name to list of values
        const flattenedProps: { [key: string]: any[] } = {};
        for (const childProfile of Object.values(profiles).filter(({ type }) => type !== baseProfileType)) {
            for (const [k, v] of Object.entries(childProfile.properties)) {
                flattenedProps[k] = [...(flattenedProps[k] || []), v];
            }
        }
        // List property names defined multiple times with the same value
        const duplicateProps: string[] = [];
        for (const [k, v] of Object.entries(flattenedProps)) {
            if (v.length > 1 && (new Set(v)).size === 1) {
                duplicateProps.push(k);
            }
        }
        // Remove duplicate properties from child profiles and store them in root profile
        for (const propName of duplicateProps) {
            profiles[baseProfileType].properties = {
                [propName]: flattenedProps[propName][0],
                ...profiles[baseProfileType].properties
            };
            for (const childProfile of Object.values(profiles).filter(({ type }) => type !== baseProfileType)) {
                delete childProfile.properties[propName];
            }
        }
    }
}
