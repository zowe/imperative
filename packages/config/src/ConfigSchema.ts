/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import { ICommandProfileProperty } from "../../cmd";
import { IProfileProperty, IProfileSchema, IProfileTypeConfiguration } from "../../profiles";
import { IConfigSchema } from "./doc/IConfigSchema";

export class ConfigSchema {
    /**
     * JSON schema URI stored in $schema property of the schema
     * @readonly
     * @memberof ConfigSchema
     */
    private static readonly JSON_SCHEMA = "https://json-schema.org/draft/2020-12/schema";

    /**
     * Version number stored in $version property of the schema
     * @readonly
     * @memberof ConfigSchema
     */
    private static readonly SCHEMA_VERSION = 3;

    /**
     * Transform an Imperative profile schema to a JSON schema. Removes any
     * non-JSON-schema properties and translates anything useful.
     * @param schema The Imperative profile schema
     * @returns JSON schema for profile properties
     */
    private static generateJsonSchema(schema: IProfileSchema): any {
        const properties: { [key: string]: any } = {};
        const secureProps: string[] = [];
        for (const [k, v] of Object.entries(schema.properties || {})) {
            properties[k] = { type: v.type };
            const cmdProp = v as ICommandProfileProperty;
            if (cmdProp.optionDefinition != null) {
                properties[k].description = cmdProp.optionDefinition.description;
                if (cmdProp.optionDefinition.defaultValue != null) {
                    properties[k].default = cmdProp.optionDefinition.defaultValue;
                }
                if (cmdProp.optionDefinition.allowableValues != null) {
                    properties[k].enum = cmdProp.optionDefinition.allowableValues.values;
                }
            }
            if (v.secure) {
                secureProps.push(k);
            }
        }

        const propertiesSchema: any = {
            type: schema.type,
            title: schema.title,
            description: schema.description,
            properties
        };
        if (schema.required) {
            propertiesSchema.required = schema.required;
        }

        const secureSchema: any = {
            prefixItems: {
                enum: secureProps
            }
        };

        if (secureProps.length > 0) {
            return { properties: propertiesSchema, secure: secureSchema };
        } else {
            return { properties: propertiesSchema };
        }
    }

    /**
     * Transform a JSON schema to an Imperative profile schema.
     * @param schema The JSON schema for profile properties
     * @returns Imperative profile schema
     */
    private static parseJsonSchema(schema: any): IProfileSchema {
        const properties: { [key: string]: IProfileProperty } = {};
        for (const [k, v] of Object.entries((schema.properties.properties || {}) as { [key: string]: any })) {
            properties[k] = { type: v.type };
            if (schema.secure?.prefixItems.enum.includes(k)) {
                properties[k].secure = true;
            }
            if (v.description != null || v.default != null || v.enum != null) {
                (properties[k] as ICommandProfileProperty).optionDefinition = {
                    name: k,
                    type: v.type,
                    description: v.description,
                    defaultValue: v.default,
                    allowableValues: v.enum ? { values: v.enum } : undefined
                };
            }
        }

        return {
            title: schema.properties.title,
            description: schema.properties.description,
            type: schema.properties.type,
            properties,
            required: schema.properties.required
        };
    }

    /**
     * Dynamically builds the config schema for this CLI.
     * @param profiles The profiles supported by this CLI
     * @returns JSON schema for all supported profile types
     */
    public static buildSchema(profiles: IProfileTypeConfiguration[]): IConfigSchema {
        const entries: any[] = [];
        const defaultProperties: { [key: string]: any } = {};
        profiles.forEach((profile: { type: string, schema: IProfileSchema }) => {
            entries.push({
                if: { properties: { type: { const: profile.type } } },
                then: { properties: this.generateJsonSchema(profile.schema) }
            });
            defaultProperties[profile.type] = { type: "string" };
        });
        return {
            $schema: ConfigSchema.JSON_SCHEMA,
            $version: ConfigSchema.SCHEMA_VERSION,
            type: "object",
            description: "config",
            properties: {
                profiles: {
                    type: "object",
                    description: "named profiles config",
                    patternProperties: {
                        "^\\S*$": {
                            type: "object",
                            description: "a profile",
                            properties: {
                                type: {
                                    description: "the profile type",
                                    type: "string"
                                },
                                properties: {
                                    description: "the profile properties",
                                    type: "object"
                                },
                                profiles: {
                                    description: "additional sub-profiles",
                                    type: "object",
                                    $ref: "#/properties/profiles"
                                },
                                secure: {
                                    description: "secure property names",
                                    type: "array",
                                    prefixItems: {
                                        type: "string"
                                    },
                                    uniqueItems: true
                                }
                            },
                            allOf: entries
                        }
                    }
                },
                defaults: {
                    type: "object",
                    description: "default profiles config",
                    properties: defaultProperties
                }
            }
        };
    }

    /**
     * Loads Imperative profile schema objects from a schema JSON file.
     * @param schemaJson The schema JSON for config
     */
    public static loadProfileSchemas(schemaJson: IConfigSchema): IProfileTypeConfiguration[] {
        const patternName = Object.keys(schemaJson.properties.profiles.patternProperties)[0];
        const profileSchemas: IProfileTypeConfiguration[] = [];
        for (const obj of schemaJson.properties.profiles.patternProperties[patternName].allOf) {
            profileSchemas.push({
                type: obj.if.properties.type.const,
                schema: this.parseJsonSchema(obj.then.properties)
            });
        }
        return profileSchemas;
    }
}
