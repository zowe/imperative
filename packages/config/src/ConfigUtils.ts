/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import { ICommandArguments } from "../../cmd";
import { ImperativeConfig } from "../../utilities";
import { ImperativeError } from "../../error";

/**
 * Coeerces string property value to a boolean or number type.
 * @param value String value
 * @returns Boolean, number, or string
 */
export function coercePropValue(value: any) {
    if (value === "true")
        return true;
    if (value === "false")
        return false;
    if (!isNaN(value) && !isNaN(parseFloat(value)))
        return parseInt(value, 10);
    return value;
}

/**
 * Retrieves the name of the active profile for the given type. If no such
 * profile exists, returns the default name which can be used to create a new profile.
 * @param profileType The type of CLI profile
 * @param cmdArguments CLI arguments which may specify a profile
 * @param defaultProfileName Name to fall back to if profile doesn't exist. If
 *                           not specified, the profile type will be used.
 * @returns The profile name
 */
export function getActiveProfileName(profileType: string, cmdArguments: ICommandArguments, defaultProfileName?: string): string {
    // Look for profile name first in command line arguments, second in
    // default profiles defined in config, and finally fall back to using
    // the profile type as the profile name.
    return cmdArguments[`${profileType}-profile`] ||
        ImperativeConfig.instance.config.properties.defaults[profileType] ||
        defaultProfileName || profileType;
}

/**
 * Form an error message for failures to securely save a value.
 *
 * @param solution Text that our caller can supply for a solution.
 */
export function secureSaveError(solution?: string): ImperativeError {
    const displayName = ImperativeConfig.instance.loadedConfig.productDisplayName || ImperativeConfig.instance.loadedConfig.name;
    let details = `Possible Solutions:\n` +
        ` 1. Reinstall ${displayName}. On Linux systems, also make sure to install the prerequisites listed in ${displayName} documentation.\n` +
        ` 2. Ensure ${displayName} can access secure credential storage. ${displayName} needs access to the OS to securely save credentials.`;
    if (solution != null) {
        details += `\n 3. ${solution}`;
    }
    return new ImperativeError({
        msg: "Unable to securely save credentials.",
        additionalDetails: details
    });
}
