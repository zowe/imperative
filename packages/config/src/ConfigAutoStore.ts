/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import * as lodash from "lodash";
import { ICommandArguments, IHandlerParameters } from "../../cmd";
import { ICommandHandlerRequire } from "../../cmd/src/doc/handler/ICommandHandlerRequire";
import { ICommandProfileAuthConfig } from "../../cmd/src/doc/profiles/definition/ICommandProfileAuthConfig";
import { IConfigLoadedProfile } from "./doc/IConfigLoadedProfile";
import * as ConfigUtils from "./ConfigUtils";
import { AbstractAuthHandler } from "../../imperative/src/auth/handlers/AbstractAuthHandler";
import { ImperativeConfig } from "../../utilities";
import { ISession } from "../../rest/src/session/doc/ISession";
import { Session } from "../../rest/src/session/Session";
import { AUTH_TYPE_TOKEN } from "../../rest/src/session/SessConstants";
import { Logger } from "../../logger";

/**
 * Class to manage automatic storage of properties in team config.
 */
export class ConfigAutoStore {
    /**
     * Finds the token auth handler class for a team config profile.
     * @param profilePath JSON path of profile
     * @param cmdArguments CLI arguments which may specify a profile
     * @returns Auth handler class or undefined if none was found
     */
    public static findAuthHandlerForProfile(profilePath: string, cmdArguments: ICommandArguments): AbstractAuthHandler | undefined {
        const config = ImperativeConfig.instance.config;
        const profileType = lodash.get(config.properties, `${profilePath}.type`);
        const profile = config.api.profiles.get(profilePath.replace(/profiles\./g, ""));

        if (profile == null || profileType == null) {  // Profile must exist and have type defined
            return;
        } else if (profileType === "base") {
            if (profile.tokenType == null) {  // Base profile must have tokenType defined
                return;
            }
        } else {
            if (profile.basePath == null) {  // Service profiles must have basePath defined
                return;
            } else if (profile.tokenType == null) {  // If tokenType undefined in service profile, fall back to base profile
                const baseProfileName = ConfigUtils.getActiveProfileName("base", cmdArguments);
                return this.findAuthHandlerForProfile(config.api.profiles.expandPath(baseProfileName), cmdArguments);
            }
        }

        const authConfigs: ICommandProfileAuthConfig[] = [];
        ImperativeConfig.instance.loadedConfig.profiles.forEach((profCfg) => {
            if ((profCfg.type === profileType || profCfg.type === "base") && profCfg.authConfig != null) {
                authConfigs.push(...profCfg.authConfig);
            }
        });

        for (const authConfig of authConfigs) {
            const authHandler: ICommandHandlerRequire = require(authConfig.handler);
            const authHandlerClass = new authHandler.default();

            if (authHandlerClass instanceof AbstractAuthHandler) {
                const promptParams = authHandlerClass.getPromptParams()[0];

                if (profile.tokenType === promptParams.defaultTokenType) {
                    return authHandlerClass;  // Auth service must have matching token type
                }
            }
        }
    }

    /**
     * Finds the highest priority layer where a profile is stored.
     * @param loadedProfile
     * @returns User and global properties
     */
    public static getPriorityLayer(loadedProfile: IConfigLoadedProfile): { user: boolean, global: boolean } {
        return {
            user: Object.values(loadedProfile.properties).every(v => v.user),
            global: Object.values(loadedProfile.properties).some(v => v.global)
        };
    }

    /**
     * Stores session config properties into a team config profile.
     * @param params CLI handler parameters object
     * @param sessCfg Session config containing properties to store
     * @param propsToStore Names of properties that should be stored
     */
    public static async storeSessCfgProps(params: IHandlerParameters, sessCfg: { [key: string]: any }, propsToStore: string[]): Promise<void> {
        const config = ImperativeConfig.instance.config;
        // TODO Which autoStore value should take priority if it conflicts between layers
        if (propsToStore.length == 0 || !config?.exists || !config.properties.autoStore) {
            return;
        }

        let profileProps = propsToStore.map(propName => propName === "hostname" ? "host" : propName);
        const profileData = this.findActiveProfile(params, profileProps);
        if (profileData == null) {
            return;
        }
        const [profileType, profileName] = profileData;
        const profilePath = config.api.profiles.expandPath(profileName);

        // Replace user and password with tokenValue if tokenType is defined in config
        if (profileProps.includes("user") && profileProps.includes("password") &&
            await this.fetchTokenForSessCfg(params, sessCfg, profilePath)) {
            profileProps = profileProps.filter(propName => propName !== "user" && propName !== "password");
            profileProps.push("tokenValue");
        }

        const beforeLayer = config.api.layers.get();
        const loadedProfile = config.api.profiles.load(profileName);
        const { user, global } = this.getPriorityLayer(loadedProfile);
        config.api.layers.activate(user, global);

        const baseProfileName = ConfigUtils.getActiveProfileName("base", params.arguments);
        const baseProfileObj = lodash.get(config.properties, config.api.profiles.expandPath(baseProfileName));
        const baseProfileSchema = ImperativeConfig.instance.loadedConfig.baseProfile.schema;
        const profileSchema = ImperativeConfig.instance.loadedConfig.profiles.find(p => p.type === profileType).schema;

        for (const propName of profileProps) {
            let propProfilePath = profilePath;
            // Determine if property should be stored in base profile instead
            if (loadedProfile.properties[propName] == null && !loadedProfile.secure?.includes(propName) &&
                (baseProfileObj.properties[propName] != null || baseProfileObj.secure?.includes(propName) ||
                (propName === "tokenValue" && baseProfileObj.properties.tokenType != null))) {
                propProfilePath = config.api.profiles.expandPath(baseProfileName);
            }
            const sessCfgPropName = propName === "host" ? "hostname" : propName;
            const propDefinition = profileSchema.properties[propName] || baseProfileSchema.properties[propName];
            config.set(`${propProfilePath}.properties.${propName}`, sessCfg[sessCfgPropName], {
                secure: propDefinition.secure
            });
        }

        await config.save(false);
        params.response.console.log(`Stored properties in ${config.layerActive().path}: ${profileProps.join(", ")}`);
        // Restore original active layer
        config.api.layers.activate(beforeLayer.user, beforeLayer.global);
    }

    /**
     * Finds the profile where auto-store properties should be saved.
     * @param params CLI handler parameters object
     * @param profileProps List of properties required in the profile schema
     * @returns Tuple containing profile type and name, or undefined if no
     *          profile was found
     */
    private static findActiveProfile(params: IHandlerParameters, profileProps: string[]): [string, string] | undefined {
        const profileTypes = [
            ...(params.definition.profile?.required || []),
            ...(params.definition.profile?.optional || [])
        ];

        for (const profType of profileTypes) {
            const profileMatch = ImperativeConfig.instance.loadedConfig.profiles.find(p => p.type === profType);
            if (profileMatch != null && profileProps.every(propName => propName in profileMatch.schema.properties)) {
                return [profType, ConfigUtils.getActiveProfileName(profType, params.arguments)];
            }
        }
    }

    /**
     * Retrieves token value that will be auto-stored into session config.
     * @param params CLI handler parameters object
     * @param sessCfg Session config with credentials for basic or cert auth
     * @param profilePath JSON path of profile containing tokenType
     * @returns True if auth handler was found and token was fetched
     */
    private static async fetchTokenForSessCfg(params: IHandlerParameters, sessCfg: { [key: string]: any }, profilePath: string): Promise<boolean> {
        const authHandlerClass = this.findAuthHandlerForProfile(profilePath, params.arguments);

        if (authHandlerClass == null) {
            return false;
        }

        const [promptParams, loginHandler] = authHandlerClass.getPromptParams();
        sessCfg.type = AUTH_TYPE_TOKEN;
        sessCfg.tokenType = promptParams.defaultTokenType;
        const baseSessCfg: ISession = { type: sessCfg.type };

        for (const propName of Object.keys(ImperativeConfig.instance.loadedConfig.baseProfile.schema.properties)) {
            const sessCfgPropName = propName === "host" ? "hostname" : propName;
            if (sessCfg[sessCfgPropName] != null) {
                (baseSessCfg as any)[sessCfgPropName] = sessCfg[sessCfgPropName];
            }
        }

        Logger.getAppLogger().info(`Fetching ${sessCfg.tokenType} for ${profilePath}`);
        sessCfg.tokenValue = await loginHandler(new Session(baseSessCfg));
        return true;
    }
}
