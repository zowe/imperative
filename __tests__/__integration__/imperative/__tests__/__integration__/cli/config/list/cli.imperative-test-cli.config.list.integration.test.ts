/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import { ITestEnvironment } from "../../../../../../../__src__/environment/doc/response/ITestEnvironment";
import { SetupTestEnvironment } from "../../../../../../../__src__/environment/SetupTestEnvironment";
import { runCliScript } from "../../../../../../../src/TestUtil";
import { expectedConfigObject } from "../__resources__/expectedObjects";
import * as path from "path";
import * as lodash from "lodash";

// Test Environment populated in the beforeAll();
let TEST_ENVIRONMENT: ITestEnvironment;

describe("imperative-test-cli config list", () => {
    let expectedGlobalProjectConfigLocation: string;
    let expectedGlobalUserConfigLocation: string;
    let expectedProjectConfigLocation: string;
    let expectedUserConfigLocation: string;
    // Create the test environment
    beforeAll(async () => {
        TEST_ENVIRONMENT = await SetupTestEnvironment.createTestEnv({
            cliHomeEnvVar: "IMPERATIVE_TEST_CLI_CLI_HOME",
            testName: "imperative_test_cli_test_config_list_command"
        });
        runCliScript(__dirname + "/../init/__scripts__/init_config.sh", TEST_ENVIRONMENT.workingDir, ["--prompt false"]);
        runCliScript(__dirname + "/../init/__scripts__/init_config.sh", TEST_ENVIRONMENT.workingDir, ["--user-config --prompt false"]);
        runCliScript(__dirname + "/../init/__scripts__/init_config.sh", TEST_ENVIRONMENT.workingDir, ["--global-config --prompt false"]);
        runCliScript(__dirname + "/../init/__scripts__/init_config.sh", TEST_ENVIRONMENT.workingDir,
            ["--user-config --global-config --prompt false"]);
        expectedGlobalUserConfigLocation = path.join(TEST_ENVIRONMENT.workingDir, "imperative-test-cli.config.user.json");
        expectedGlobalProjectConfigLocation = path.join(TEST_ENVIRONMENT.workingDir, "imperative-test-cli.config.json");
        expectedUserConfigLocation = path.join(TEST_ENVIRONMENT.workingDir, "test", "imperative-test-cli.config.user.json");
        expectedProjectConfigLocation = path.join(TEST_ENVIRONMENT.workingDir, "test", "imperative-test-cli.config.json");
    });
    afterAll(() => {
        runCliScript(__dirname + "/../__scripts__/delete_configs.sh", TEST_ENVIRONMENT.workingDir,
            ["-rf imperative-test-cli.config.user.json imperative-test-cli.config.json test imperative-test-cli.schema.json"]);
    });
    it("should display the help", () => {
        const response = runCliScript(__dirname + "/../__scripts__/get_help.sh",
            TEST_ENVIRONMENT.workingDir, ["list"]);
        expect(response.stdout.toString()).toContain(`List config properties`);
        expect(response.stderr.toString()).toEqual("");
        expect(response.error).not.toBeDefined();
    });
    it("should list the configuration", () => {
        const response = runCliScript(__dirname + "/__scripts__/list_config.sh", TEST_ENVIRONMENT.workingDir, [""]);
        expect(response.stdout.toString()).toMatchSnapshot();
        expect(response.stdout.toString()).toContain("secured: my_profiles.secured");
        expect(response.stdout.toString()).toContain("type:       secured");
        expect(response.stdout.toString()).toContain("defaults:");
        expect(response.stdout.toString()).toContain("profiles:");
        expect(response.stdout.toString()).toContain("secure:");
        expect(response.stdout.toString()).toContain("(empty array)");
        expect(response.stderr.toString()).toEqual("");
        expect(response.error).not.toBeDefined();
    });
    it("should list the configuration in RFJ", () => {
        const response = runCliScript(__dirname + "/__scripts__/list_config.sh", TEST_ENVIRONMENT.workingDir, ["--rfj"]);
        const parsedResponse = JSON.parse(response.stdout.toString());
        const expectedResponse = {
            data: {
                defaults: {
                    secured: "my_profiles.secured",
                    base: "my_base"
                },
                profiles: {
                    my_base: {
                        properties: {},
                        type: "base",
                        secure: []
                    },
                    my_profiles: {
                        profiles: {
                            secured: {
                                type: "secured",
                                properties: {
                                    info: ""
                                },
                                secure: [
                                    "secret"
                                ]
                            }
                        },
                        properties: {}
                    }
                },
                overrides: {}
            }
        };
        expect(parsedResponse.success).toEqual(true);
        expect(parsedResponse.stderr).toEqual("");
        expect(parsedResponse.exitCode).toEqual(0);
        expect(parsedResponse.data).toEqual(expectedResponse.data);
    });
    it("should list the configurations based on location", () => {
        const response = runCliScript(__dirname + "/__scripts__/list_config.sh", TEST_ENVIRONMENT.workingDir, ["--locations"]);
        expect(response.stdout.toString()).toContain(expectedProjectConfigLocation);
        expect(response.stdout.toString()).toContain(expectedUserConfigLocation);
        expect(response.stdout.toString()).toContain(expectedGlobalProjectConfigLocation);
        expect(response.stdout.toString()).toContain(expectedGlobalUserConfigLocation);
        expect(response.stdout.toString()).toContain("defaults:");
        expect(response.stdout.toString()).toContain("profiles:");
        expect(response.stdout.toString()).toContain("secure:");
        expect(response.stdout.toString()).toContain("my_profiles:");
        expect(response.stdout.toString()).toMatch(/type:\s+secured/);
        expect(response.stdout.toString()).toContain("properties:");
        expect(response.stdout.toString()).toMatch(/secured:\s+my_profiles\.secured/);
        expect(response.stdout.toString()).toMatch(/\$schema:\s+\.\/imperative-test-cli\.schema\.json/);
        expect(response.stderr.toString()).toEqual("");
        expect(response.error).not.toBeDefined();
    });
    it("should list the configurations based on location in RFJ", () => {
        const response = runCliScript(__dirname + "/__scripts__/list_config.sh", TEST_ENVIRONMENT.workingDir, ["--locations --rfj"]);
        const parsedResponse = JSON.parse(response.stdout.toString());
        const expectedUserConfig = {
            $schema: "./imperative-test-cli.schema.json",
            profiles: {
                my_base: {
                    properties: {},
                    type: "base",
                    secure: []
                },
                my_profiles: {
                    profiles: {
                        secured: {
                            properties: {},
                            type: "secured",
                            secure: []
                        }
                    },
                    properties: {}
                }
            },
            defaults: {},
            overrides: {}
        };
        const expectedProjectConfig = lodash.cloneDeep(expectedConfigObject);
        const expectedResponse = {
            data: {}
        };
        expectedResponse.data[expectedUserConfigLocation] = expectedUserConfig;
        expectedResponse.data[expectedGlobalUserConfigLocation] = expectedUserConfig;
        expectedResponse.data[expectedGlobalProjectConfigLocation] = expectedProjectConfig;
        expectedResponse.data[expectedProjectConfigLocation] = expectedProjectConfig;
        expect(parsedResponse.success).toEqual(true);
        expect(parsedResponse.stderr).toEqual("");
        expect(parsedResponse.exitCode).toEqual(0);
        expect(parsedResponse.data).toEqual(expectedResponse.data);
    });
    it("should list the root level property names only", () => {
        const response = runCliScript(__dirname + "/__scripts__/list_config.sh", TEST_ENVIRONMENT.workingDir, ["--root"]);
        expect(response.stdout.toString()).toMatchSnapshot();
        expect(response.stdout.toString()).toContain("defaults");
        expect(response.stdout.toString()).toContain("profiles");
        expect(response.stderr.toString()).toEqual("");
        expect(response.error).not.toBeDefined();
    });
    it("should get a list of config file paths", () => {
        const response = runCliScript(__dirname + "/__scripts__/list_config.sh", TEST_ENVIRONMENT.workingDir, ["--locations --root"]);
        expect(response.stdout.toString()).toContain(expectedProjectConfigLocation);
        expect(response.stdout.toString()).toContain(expectedUserConfigLocation);
        expect(response.stdout.toString()).toContain(expectedGlobalProjectConfigLocation);
        expect(response.stdout.toString()).toContain(expectedGlobalUserConfigLocation);
        expect(response.stderr.toString()).toEqual("");
        expect(response.error).not.toBeDefined();
    });
    it("should list the profiles configuration property", () => {
        const response = runCliScript(__dirname + "/__scripts__/list_config.sh", TEST_ENVIRONMENT.workingDir, ["profiles"]);
        expect(response.stdout.toString()).toMatchSnapshot();
        expect(response.stdout.toString()).toContain("my_profiles:");
        expect(response.stdout.toString()).toContain("secured:");
        expect(response.stdout.toString()).toContain("type:       secured");
        expect(response.stdout.toString()).toContain("properties:");
        expect(response.stderr.toString()).toEqual("");
        expect(response.error).not.toBeDefined();
    });
    it("should list the defaults configuration property", () => {
        const response = runCliScript(__dirname + "/__scripts__/list_config.sh", TEST_ENVIRONMENT.workingDir, ["defaults"]);
        expect(response.stdout.toString()).toMatchSnapshot();
        expect(response.stdout.toString()).toContain("secured: my_profiles.secured");
        expect(response.stderr.toString()).toEqual("");
        expect(response.error).not.toBeDefined();
    });
    it("should list the configuration without showing secure values", () => {
        runCliScript(__dirname + "/../set/__scripts__/set_secure.sh", TEST_ENVIRONMENT.workingDir, ["profiles.my_profiles.profiles.secured.properties.secret", "area51"]);
        const response = runCliScript(__dirname + "/__scripts__/list_config.sh", TEST_ENVIRONMENT.workingDir, [""]);
        expect(response.stdout.toString()).toMatchSnapshot();
        expect(response.stdout.toString()).toContain("secured: my_profiles.secured");
        expect(response.stdout.toString()).toContain("type:       secured");
        expect(response.stdout.toString()).toContain("defaults:");
        expect(response.stdout.toString()).toContain("profiles:");
        expect(response.stdout.toString()).toContain("secure:");
        expect(response.stdout.toString()).toContain("(empty array)");
        expect(response.stdout.toString()).toContain("secret: (secure value)");
        expect(response.stderr.toString()).toEqual("");
        expect(response.error).not.toBeDefined();
    });
});
