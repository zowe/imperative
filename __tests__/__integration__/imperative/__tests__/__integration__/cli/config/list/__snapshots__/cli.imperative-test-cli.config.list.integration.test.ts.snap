// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`imperative-test-cli config list should list the configuration 1`] = `
"profiles: 
  secured: 
    type:       secured
    properties: 
      info: 
    secure: 
      - secret
  base: 
    type:       base
    properties: 
    secure: 
      (empty array)
defaults: 
  secured: secured
  base:    base
"
`;

exports[`imperative-test-cli config list should list the configuration without showing secure values 1`] = `
"profiles: 
  secured: 
    type:       secured
    properties: 
      info:   
      secret: (secure value)
    secure: 
      - secret
  base: 
    type:       base
    properties: 
    secure: 
      (empty array)
defaults: 
  secured: secured
  base:    base
"
`;

exports[`imperative-test-cli config list should list the defaults configuration property 1`] = `
"secured: secured
base:    base
"
`;

exports[`imperative-test-cli config list should list the profiles configuration property 1`] = `
"secured: 
  type:       secured
  properties: 
    info: 
  secure: 
    - secret
base: 
  type:       base
  properties: 
  secure: 
    (empty array)
"
`;

exports[`imperative-test-cli config list should list the root level property names only 1`] = `
"profiles
defaults
"
`;
